File: packages/vm/wasmlib/go/wasmclient/service.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmclient

import (
	"fmt"
	"strings"
	"time"

	"github.com/iotaledger/hive.go/crypto/ed25519"
	"github.com/iotaledger/wasp/client"
	"github.com/iotaledger/wasp/packages/iscp"
	"github.com/iotaledger/wasp/packages/iscp/colored"
	"github.com/iotaledger/wasp/packages/iscp/request"
	"github.com/iotaledger/wasp/packages/iscp/requestargs"
	"github.com/iotaledger/wasp/packages/kv"
	"github.com/iotaledger/wasp/packages/kv/dict"
	"github.com/iotaledger/wasp/packages/subscribe"
	"github.com/mr-tron/base58"
)

type ArgMap dict.Dict

func (m ArgMap) Get(key string) []byte {
	return m[kv.Key(key)]
}

func (m ArgMap) Set(key string, value []byte) {
	m[kv.Key(key)] = value
}

type ResMap dict.Dict

func (m ResMap) Get(key string) []byte {
	return m[kv.Key(key)]
}

type IEventHandler interface {
	CallHandler(topic string, params []string)
}

type Service struct {
	chainID       *iscp.ChainID
	eventHandlers []IEventHandler
	keyPair       *ed25519.KeyPair
	scHname       iscp.Hname
	waspClient    *client.WaspClient
}

func (s *Service) Init(svcClient *ServiceClient, chainID string, scHname uint32) (err error) {
	s.waspClient = svcClient.waspClient
	s.scHname = iscp.Hname(scHname)
	s.chainID, err = iscp.ChainIDFromString(chainID)
	if err != nil {
		return err
	}
	return s.startEventHandlers(svcClient.eventPort)
}

func (s *Service) AsClientFunc() ClientFunc {
	return ClientFunc{svc: s}
}

func (s *Service) AsClientView() ClientView {
	return ClientView{svc: s}
}

func (s *Service) CallView(viewName string, args ArgMap) (ResMap, error) {
	res, err := s.waspClient.CallView(s.chainID, s.scHname, viewName, dict.Dict(args))
	if err != nil {
		return nil, err
	}
	return ResMap(res), nil
}

func (s *Service) PostRequest(hFuncName uint32, args ArgMap, transfer *Transfer, keyPair *ed25519.KeyPair, onLedger bool) Request {
	bal, err := makeBalances(transfer)
	if err != nil {
		return Request{err: err}
	}
	reqArgs := requestargs.New()
	if args != nil {
		reqArgs.AddEncodeSimpleMany(dict.Dict(args))
	}

	if onLedger {
		return s.postRequestOnLedger(hFuncName, reqArgs, bal, keyPair)
	}

	req := request.NewOffLedger(s.chainID, s.scHname, iscp.Hname(hFuncName), reqArgs)
	req.WithTransfer(bal)
	req.Sign(keyPair)
	err = s.waspClient.PostOffLedgerRequest(s.chainID, req)
	if err != nil {
		return Request{err: err}
	}
	id := req.ID()
	return Request{id: &id}
}

func (s *Service) postRequestOnLedger(hFuncName uint32, args requestargs.RequestArgs, bal colored.Balances, pair *ed25519.KeyPair) Request {
	// TODO implement
	return Request{}
}

func (s *Service) Register(handler IEventHandler) {
	for _, h := range s.eventHandlers {
		if h == handler {
			return
		}
	}
	s.eventHandlers = append(s.eventHandlers, handler)
}

// overrides default contract name
func (s *Service) ServiceContractName(contractName string) {
	s.scHname = iscp.Hn(contractName)
}

func (s *Service) SignRequests(keyPair *ed25519.KeyPair) {
	s.keyPair = keyPair
}

func (s *Service) Unegister(handler IEventHandler) {
	for i, h := range s.eventHandlers {
		if h == handler {
			s.eventHandlers = append(s.eventHandlers[:i], s.eventHandlers[i+1:]...)
			return
		}
	}
}

func (s *Service) WaitRequest(req Request) error {
	return s.waspClient.WaitUntilRequestProcessed(s.chainID, *req.id, 1*time.Minute)
}

func (s *Service) startEventHandlers(eventPort string) error {
	chMsg := make(chan []string, 20)
	chDone := make(chan bool)
	err := subscribe.Subscribe(eventPort, chMsg, chDone, true, "")
	if err != nil {
		return err
	}
	go func() {
		for {
			for msgSplit := range chMsg {
				event := strings.Join(msgSplit, " ")
				fmt.Printf("%s\n", event)
				if msgSplit[0] == "vmmsg" {
					msg := strings.Split(msgSplit[3], "|")
					topic := msg[0]
					params := msg[1:]
					for _, handler := range s.eventHandlers {
						handler.CallHandler(topic, params)
					}
				}
			}
		}
	}()
	return nil
}

/////////////////////////////////////////////////////////////////

func Base58Decode(s string) []byte {
	res, err := base58.Decode(s)
	if err != nil {
		panic("invalid base58 encoding")
	}
	return res
}

func Base58Encode(b []byte) string {
	return base58.Encode(b)
}

func makeBalances(transfer *Transfer) (colored.Balances, error) {
	cb := colored.NewBalances()
	if transfer != nil {
		for color, amount := range transfer.xfer {
			c, err := colored.ColorFromBase58EncodedString(color)
			if err != nil {
				return nil, err
			}
			cb.Set(c, amount)
		}
	}
	return cb, nil
}

File: packages/vm/wasmlib/ts/wasmclient/service.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import * as wasmclient from "./index";
import {Hash, IKeyPair} from "./crypto";
import {IOnLedger} from "./goshimmer/models/on_ledger";
import {Colors} from "./colors";
import {Buffer} from './buffer';

export interface IEventHandler {
    callHandler(topic: string, params: string[]): void;
}

export class Service {
    private serviceClient: wasmclient.ServiceClient;
    private webSocket: WebSocket | null = null;
    public keyPair: IKeyPair | null = null;
    private eventHandlers: Array<IEventHandler> = new Array();
    public scHname: wasmclient.Hname;
    private waspWebSocketUrl: string = "";

    constructor(client: wasmclient.ServiceClient, scHname: wasmclient.Hname) {
        this.serviceClient = client;
        this.scHname = scHname;
        this.configureWebSocketsEventHandlers();
    }

    public async callView(viewName: string, args: wasmclient.Arguments, res: wasmclient.Results): Promise<void> {
        await this.serviceClient.waspClient.callView(
            this.serviceClient.configuration.chainId,
            this.scHname.toString(16),
            viewName,
            args.encodeCall(),
            res
        );
    }

    public async postRequest(
        hFuncName: wasmclient.Int32,
        args: wasmclient.Arguments,
        transfer: wasmclient.Transfer,
        keyPair: IKeyPair,
        onLedger: boolean
    ): Promise<string> {
        const chainId = this.serviceClient.configuration.chainId;
        if (! onLedger) {
            // requested off-ledger request
            const requestID = await this.serviceClient.waspClient.postOffLedgerRequest(chainId, this.scHname, hFuncName, args, transfer, keyPair);
            return requestID;
        }

        // requested on-ledger request
        const payload: IOnLedger = {
            contract: this.scHname,
            entrypoint: hFuncName,
            //TODO: map args
            //arguments : args
        };
        const transferAmount = transfer.get(Colors.IOTA_COLOR);
        const transactionID = await this.serviceClient.goShimmerClient.postOnLedgerRequest(chainId, payload, transferAmount, keyPair);
        if (!transactionID) throw new Error("No transaction id");
        return transactionID;
    }

    public register(handler: IEventHandler): void {
        for (let i = 0; i < this.eventHandlers.length; i++) {
            if (this.eventHandlers[i] === handler) {
                return;
            }
        }
        this.eventHandlers.push(handler);
    }

    public unregister(handler: IEventHandler): void {
        // remove handler
        this.eventHandlers = this.eventHandlers.filter(h => h !== handler);
    }

    // overrides default contract name
    public serviceContractName(contractName: string): void {
        this.scHname = Hash.from(Buffer.from(contractName)).readUInt32LE(0)
    }

    public async waitRequest(reqID: wasmclient.RequestID): Promise<void> {
        await this.serviceClient.waspClient.waitRequest(this.serviceClient.configuration.chainId, reqID);
    }

    private configureWebSocketsEventHandlers() {
        if (
            this.serviceClient.configuration.waspWebSocketUrl.startsWith("wss://") ||
            this.serviceClient.configuration.waspWebSocketUrl.startsWith("ws://")
        )
            this.waspWebSocketUrl = this.serviceClient.configuration.waspWebSocketUrl;
        else this.waspWebSocketUrl = "ws://" + this.serviceClient.configuration.waspWebSocketUrl;

        this.waspWebSocketUrl = this.waspWebSocketUrl.replace("%chainId", this.serviceClient.configuration.chainId);

        this.connectWebSocket();
    }

    private connectWebSocket(): void {
        // eslint-disable-next-line no-console
        console.log(`Connecting to Websocket => ${this.waspWebSocketUrl}`);
        this.webSocket = new WebSocket(this.waspWebSocketUrl);
        this.webSocket.addEventListener("message", (x) => this.handleIncomingMessage(x));
        this.webSocket.addEventListener("close", () => setTimeout(this.connectWebSocket.bind(this), 1000));
    }

    private handleIncomingMessage(message: MessageEvent<string>): void {
        // expect vmmsg <chain ID> <contract hname> contract.event|parameters
        const msg = message.data.toString().split(" ");
        if (msg.length != 4 || msg[0] != "vmmsg") {
            return;
        }
        const topics = msg[3].split("|");
        const topic = topics[0];
        const params = topics.slice(1);
        for (let i = 0; i < this.eventHandlers.length; i++) {
            this.eventHandlers[i].callHandler(topic, params);
        }
    }
}
